### PR TITLE
Support nullable variance in interface and partial method implementation.

### LINKED ISSUE #35227: Support nullable variance in interface implementation
Essentially, the work from  #23268 and #30958, but for interface implementation.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Closes #35227.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@agocke, @dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs [hunk 1]
```diff
@@ -1494,60 +1494,128 @@ private static void ReportImplicitImplementationMatchDiagnostics(Symbol interfac
...
+                                                                                                            new FormattedSymbol(implementedEvent, SymbolDisplayFormat.MinimallyQualifiedFormat));
+                                                                                        }
+                                                                                    },
+                                                                                    (implementingType, isExplicit));
                 }
                 else
                 {
-                    implementedMember = interfaceMember;
-                }
+                    Action<DiagnosticBag, MethodSymbol, MethodSymbol, (TypeSymbol implementingType, bool isExplicit)> reportMismatchInReturnType =
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs]: Consider making these local functions so we can make them static and easily verify that they are not capturing.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs [hunk 2]
```diff
@@ -875,23 +876,28 @@ private void CheckNewModifier(Symbol symbol, bool isNew, DiagnosticBag diagnosti
...
+                                        overriddenGetMethod,
                                         overridingProperty.GetMethod,
                                         diagnostics,
+                                        checkReturnType: true,
                                         // Don't check parameters on the getter if there is a setter
                                         // because they will be a subset of the setter
-                                        checkParameters: overridingProperty.SetMethod is null);
+                                        checkParameters: overridingProperty.SetMethod is null ||
+                                                         overriddenGetMethod?.AssociatedSymbol != overriddenProperty ||
+                                                         overriddenProperty.GetOwnOrInheritedSetMethod()?.AssociatedSymbol != overriddenProperty);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: Could consider extracting this check to a helper, as we use it in multiple places.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs [hunk 3]
```diff
@@ -1030,53 +1034,71 @@ static bool isOrContainsErrorType(TypeSymbol typeSymbol)
...
-                    compilation is null ||
-                    !compilation.IsFeatureEnabled(MessageID.IDS_FeatureNullableReferenceTypes))
-                {
-                    // Don't do any validation if the nullable feature is not enabled or
-                    // the override is not written directly in source
-                    return;
-                }
+                CheckValidNullableMethodOverride(overridingMethod.DeclaringCompilation, overriddenMethod, overridingMethod, diagnostics,
+                                                 checkReturnType ?
+                                                    (diagnostics, overriddenMethod, overridingMethod, location) => diagnostics.Add(ErrorCode.WRN_NullabilityMismatchInReturnTypeOnOverride, location) :
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: >overriddenMethod, overridingMethod [](start = 66, length = 34)

Consider using `_` for ignored parameter.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs [hunk 4]
```diff
@@ -1030,53 +1034,71 @@ static bool isOrContainsErrorType(TypeSymbol typeSymbol)
...
+                                                    (Action<DiagnosticBag, MethodSymbol, MethodSymbol, Location>)null,
+                                                 checkParameters ?
+                                                     (diagnostics, overriddenMethod, overridingMethod, overridingParameter, location) =>
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: >(diagnostics, overriddenMethod, overridingMethod, overridingParameter, location) => [](start = 53, length = 83)

nit: consider using a local function for those funcs in various invocations of `CheckValidNullableEventOverride`

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs [hunk 5]
```diff
@@ -1030,53 +1034,71 @@ static bool isOrContainsErrorType(TypeSymbol typeSymbol)
...
+                    conversions,
+                    overridingMethod.RefKind,
+                    overridingMethod.ReturnTypeWithAnnotations,
+                    overriddenMethod.ReturnTypeWithAnnotations))
+            {
+                reportMismatchInReturnType(diagnostics, overriddenMethod, overridingMethod, extraArgument);
+                return;
+            }
+
+            if (reportMismatchInParameterType == null)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: Consider moving this `if` before the previous one, and removing the `reportMismatchInParameterType != null` check from that one.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: Wait, ignore. Misread parameter for return.

---
In reply to: [296389113](https://github.com/dotnet/roslyn/pull/36663#discussion_r296389113) [](ancestors = 296389113)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs [hunk 6]
```diff
@@ -1106,6 +1128,44 @@ static bool isOrContainsErrorType(TypeSymbol typeSymbol)
             }
         }
 
+        private static bool CheckValidNullableOverride(
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: >CheckValidNullableOverride [](start = 28, length = 26)

nit: name of the method is a bit misleading as it doesn't actually perform a check. Maybe `ShouldCheck...`?

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs]: +1 on that

---
In reply to: [296380417](https://github.com/dotnet/roslyn/pull/36663#discussion_r296380417) [](ancestors = 296380417)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs [hunk 7]
```diff
@@ -1106,6 +1128,44 @@ static bool isOrContainsErrorType(TypeSymbol typeSymbol)
...
+            CSharpCompilation compilation,
+            Symbol overriddenMember,
+            Symbol overridingMember)
+        {
+            if (overriddenMember is null ||
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: Nit: you could just return ! this condition.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs]: Delegates for lambdas are cached automatically.

---
In reply to: [296389943](https://github.com/dotnet/roslyn/pull/36663#discussion_r296389943) [](ancestors = 296389943,296380417)

**Comment by 333fred** [comment]
Done review pass (commit 2). Mostly looks good, just a couple of small formatting suggestions.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: > Could consider extracting this check to a helper, as we use it in multiple places.

I do not believe it is worth doing just for portion of the condition, the entire condition is different.

---
In reply to: [296387584](https://github.com/dotnet/roslyn/pull/36663#discussion_r296387584) [](ancestors = 296387584)

**Comment by jcouv** [comment]
>     public abstract event System.Action<string> E1; 

If not covered elsewhere, consider extending this test to add `event Action<string?>` in the base which is overridden with `event Action<string>`. Should produce no warning.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:11173 in aff03fd. [](commit_id = aff03fd0a2c49017222a1df80e3594bd2a0dcb66, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: > nit: consider using a local function for those funcs in various invocations of CheckValidNullableEventOverride

Lambdas are used intentionally to take advantage of automatic caching.

---
In reply to: [296389099](https://github.com/dotnet/roslyn/pull/36663#discussion_r296389099) [](ancestors = 296389099)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: > Consider using _ for ignored parameter.

I see no point in doing this, it is not hard for me to type the names and it is better for readability in my opinion. Also, I do not think both parameters can be named ```_```. 

---
In reply to: [296388073](https://github.com/dotnet/roslyn/pull/36663#discussion_r296388073) [](ancestors = 296388073)

**Comment by AlekseyTs** [comment]
>     public abstract event System.Action<string> E1; 

> If not covered elsewhere, consider extending this test to add event Action<string?> in the base which is overridden with event Action<string>. Should produce no warning.

I am find existing coverage sufficient.

---
In reply to: [504570605](https://github.com/dotnet/roslyn/pull/36663#issuecomment-504570605) [](ancestors = 504570605)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:11173 in aff03fd. [](commit_id = aff03fd0a2c49017222a1df80e3594bd2a0dcb66, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs]: FWIW, for me it's less about caching and more about readability.

---
In reply to: [296390411](https://github.com/dotnet/roslyn/pull/36663#discussion_r296390411) [](ancestors = 296390411,296389943,296380417)

**Pull request merged**